### PR TITLE
Handle incoming emails

### PR DESCRIPTION
**Pull request by bengolder**

### CODE REVIEW

File: app/assets/stylesheets/components/_contact-record.scss [hunk 1]
```diff
@@ -1,5 +1,12 @@
 .contact-record {
   margin-bottom: 2rem;
+
+  &--incoming_text_message, &--outgoing_text_message {
+    .contact-record__body {
+      border-radius: 2rem;
+    }
+  }
```
Comment (bengolder) [inline app/assets/stylesheets/components/_contact-record.scss:8]: Just wanted to distinguish sms from email so I could see it easily

File: app/controllers/mailgun_webhooks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+class MailgunWebhooksController < ApplicationController
+  skip_before_action :redirect_to_getyourrefund
+  skip_before_action :set_visitor_id
+  skip_before_action :set_source
+  skip_before_action :set_referrer
+  skip_before_action :set_utm_state
+  skip_before_action :set_sentry_context
+  skip_before_action :check_maintenance_mode
+  skip_after_action :track_page_view
+  skip_before_action :verify_authenticity_token
```
Comment (bengolder) [inline app/controllers/mailgun_webhooks_controller.rb:12]: would love to refactor this inheritance and all these before_actions. Also present in TwilioWebhooksController, and would be useful in ZendeskWebhooks

File: app/models/incoming_email.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+class IncomingEmail < ApplicationRecord
+  belongs_to :client
+
+  def contact_record_type
+    self.class.name.underscore.to_sym
+  end
+
+  def body
+    stripped_html&.html_safe || stripped_text || body_html&.html_safe || body_plain
+  end
```
Comment (bengolder) [inline app/models/incoming_email.rb:38]: this is a little dangerous. Not sure how to better display html content.

File: spec/controllers/clients_controller_spec.rb [hunk 4]
```diff
@@ -118,10 +118,11 @@
...
           [
-            create(:outgoing_text_message, body: "Your tax return is great", sent_at: DateTime.new(2020, 1, 1, 0, 0, 1), client: client, twilio_status: twilio_status),
-            create(:incoming_text_message, body: "Thx appreciate yr gratitude", received_at: DateTime.new(2020, 1, 1, 0, 0, 2), client: client),
+            create(:incoming_email, body_plain: "Me too! Happy to get every notification", received_at: DateTime.new(2020, 1, 1, 0, 0, 4), client: client),
             create(:outgoing_email, body: "We are really excited to work with you", sent_at: DateTime.new(2020, 1, 1, 0, 0, 3), client: client),
-          ]
+            create(:incoming_text_message, body: "Thx appreciate yr gratitude", received_at: DateTime.new(2020, 1, 1, 0, 0, 2), client: client),
+            create(:outgoing_text_message, body: "Your tax return is great", sent_at: DateTime.new(2020, 1, 1, 0, 0, 1), client: client, twilio_status: twilio_status),
+          ].reverse
         end
```
Comment (bengolder) [inline spec/controllers/clients_controller_spec.rb:126]: I created these in reverse order to make sure it was sorting by datetime

File: app/models/incoming_email.rb (hunk 3 above)
Comment (bengolder) [reply app/models/incoming_email.rb:38]: should I only do plaintext for now?

File: app/controllers/mailgun_webhooks_controller.rb (hunk 2 above)
Comment (bengolder) [reply app/controllers/mailgun_webhooks_controller.rb:10]: Should we?
1. do a base WebhookController that skips all the junk
2. have webhook controllers inherit from `ActionController::Base`
3. make a new base controller for all non-webhook controllers that does all these before actions

File: app/models/concerns/contact_record.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+module ContactRecord
+  extend ActiveSupport::Concern
```
Comment (bengolder) [inline app/models/concerns/contact_record.rb:2]: this line isn't strictly necessary, but adds functionality that could be useful in the future

File: app/controllers/mailgun_webhooks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  before_action :valid_mailgun_post?
+
+  def create_incoming_email
+    sender_email = params["sender"]
```
Comment (alaroia-cfa) [inline app/controllers/mailgun_webhooks_controller.rb:14]: Is this the `Sender: ` email header? Or is it something computed by Mailgun? Or is it the "Envelope From"?

I think the "From:" header is a better here, but happy to discuss.

File: app/models/concerns/contact_record.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+module ContactRecord
```
Comment (bengolder) [inline app/models/concerns/contact_record.rb:1]: I got tired of naming mixins with the "Something-able" pattern

File: app/models/incoming_email.rb (hunk 3 above)
Comment (alaroia-cfa) [reply app/models/incoming_email.rb:37]: I vote just plaintext for now.

The safety depends how we render it in the template. If we're unsure, we can do just plaintext for now and come back to it imho.

File: app/controllers/clients_controller.rb [hunk 8]
```diff
@@ -19,8 +19,12 @@ def create
...
   def show
     @client = Client.find(params[:id])
-    @contact_history = (@client.outgoing_text_messages + @client.incoming_text_messages + @client.outgoing_emails
-                       ).sort_by(&:datetime)
+    @contact_history = (
+      @client.outgoing_text_messages.includes(:user) +
+      @client.incoming_text_messages +
+      @client.outgoing_emails.includes(:user) +
+      @client.incoming_emails
+    ).sort_by(&:datetime)
```
Comment (bengolder) [inline app/controllers/clients_controller.rb:27]: this will prefetch related user objects

File: app/services/mailgun_service.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
+class MailgunService
+  class << self
+    def valid_post?(params)
+      signing_key = EnvironmentCredentials.dig(:mailgun, :webhook_signing_key)
```
Comment (alaroia-cfa) [inline app/services/mailgun_service.rb:4]: It'd be nice to have comments about which parts of this were specifically recommended by Mailgun, or which parts we chose for our reasons (and a link to their doc)

File: app/controllers/mailgun_webhooks_controller.rb (hunk 6 above)
Comment (alaroia-cfa) [reply app/controllers/mailgun_webhooks_controller.rb:14]: Non-blocking!

**Review by alaroia-cfa — APPROVED** [review]

File: app/controllers/mailgun_webhooks_controller.rb (hunk 2 above)
Comment (alaroia-cfa) [reply app/controllers/mailgun_webhooks_controller.rb:10]: Jenny says 1 or 2, preferably 2

I say 4 or 2, preferably 4, but 2 is also great

sounds like 2 :P

File: app/controllers/mailgun_webhooks_controller.rb (hunk 6 above)
Comment (bengolder) [reply app/controllers/mailgun_webhooks_controller.rb:14]: They document it here: https://documentation.mailgun.com/en/latest/user_manual.html#routes

> sender of the message as reported by `MAIL FROM` during SMTP chat. Note: this value may differ from `From MIME header.

Comment (alaroia-cfa) [reply app/controllers/mailgun_webhooks_controller.rb:14]: decided, we'll go with sender for now

Comment (alaroia-cfa) [reply app/controllers/mailgun_webhooks_controller.rb:14]: comment with docs is nice

**Pull request merged**